[Analysis][SVE] Improve cost model for some extending masked loads

david-arm opened the pull request:
When performing a masked load of an unpacked SVE vector type, i.e.
nxv8i8, followed by a zero- or sign-extend to an illegal wide type
such as nxv8i32 we typically end up with a combination of an
extending masked load and pair(s) of uunpklo/hi or sunpklo/hi
instructions. For example, see test @masked_sload_8i8_8i32 in file

  CodeGen/AArch64/sve-masked-ldst-sext.ll

where

  %aval = call <vscale x 8 x i8> @llvm.masked.load.nxv8i8(...
  %aext = sext <vscale x 8 x i8> %aval to <vscale x 8 x i32>

gets lowered to

  ld1sb { z1.h }, ...
  sunpklo z0.s, z1.h
  sunpkhi z1.s, z1.h

Currently the cost for the 'sext' operation in the example above is
1, whereas this patch changes it to 2 to reflect the pair of
instructions required. Similarly, when doing a masked load of a
nxv8i8 and extending to nxv8i64 the cost is changed to 6 to reflect
the 6 unpacks required.

david-arm committed:
[NFC][Analysis] Add more cost tests for unusual SVE vector extends

david-arm committed:
[Analysis][SVE] Improve cost model for some extending masked loads

When performing a masked load of an unpacked SVE vector type, i.e.
nxv8i8, followed by a zero- or sign-extend to an illegal wide type
such as nxv8i32 we typically end up with a combination of an
extending masked load and pair(s) of uunpklo/hi or sunpklo/hi
instructions. For example, see test @masked_sload_8i8_8i32 in file

  CodeGen/AArch64/sve-masked-ldst-sext.ll

where

  %aval = call <vscale x 8 x i8> @llvm.masked.load.nxv8i8(...
  %aext = sext <vscale x 8 x i8> %aval to <vscale x 8 x i32>

gets lowered to

  ld1sb { z1.h }, ...
  sunpklo z0.s, z1.h
  sunpkhi z1.s, z1.h

Currently the cost for the 'sext' operation in the example above is
1, whereas this patch changes it to 2 to reflect the pair of
instructions required. Similarly, when doing a masked load of a
nxv8i8 and extending to nxv8i64 the cost is changed to 6 to reflect
the 6 unpacks required.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2461,6 +2461,34 @@ InstructionCost AArch64TTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
       return AdjustCost(Entry->Cost);
 
+  if ((ISD == ISD::ZERO_EXTEND || ISD == ISD::SIGN_EXTEND) &&
+      CCH == TTI::CastContextHint::Masked && ST->hasSVEorSME() &&
```
That sounds like the `unmasked load` is the special case here, not the masked load. For the unmasked load, SelectionDAG folds the extends into the loads entirely. For any other operation it just expands to unpacks. See https://godbolt.org/z/oEPKK5We8

david-arm replied:
Maybe that's a good way of looking at it, but right now the costs for normal load + sext/zext already look sensible, e.g. see `@scalable_ext_loads` in llvm/test/Analysis/CostModel/AArch64/sve-ldst.ll:

```
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %load2.nxv8i8.3 = load <vscale x 8 x i8>, ptr undef, align 8
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %sext.nxv8i8to64 = sext <vscale x 8 x i8> %load2.nxv8i8.3 to <vscale x 8 x i64>
```

where the cost of 3 matches up with the extra 3 extending loads after splitting. Whereas for the masked load case the extend requires an extra 6 instructions. In my patch I originally also dealt with the normal load + zext/sext explicitly too, but then I realised I'd added costs that were exactly the same as what we already get using the BasicTTIImpl code! So I changed the patch to just deal with the masked loads only.

dtemirbulatov commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2461,6 +2461,34 @@ InstructionCost AArch64TTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
 
+  if ((ISD == ISD::ZERO_EXTEND || ISD == ISD::SIGN_EXTEND) &&
+      CCH == TTI::CastContextHint::Masked && ST->hasSVEorSME() &&
+      !TLI->isTypeLegal(SrcTy) && !TLI->isTypeLegal(DstTy)) {
```
Is this change might affect non-load extend instructions?

david-arm replied:
That's a good question @dtemirbulatov! For non-load extends the context hint (CCH) will be TTI::CastContextHint::None, whereas I am checking for TTI::CastContextHint::Masked. That means the new code I've added only applies when we are extending a masked load.

david-arm committed:
Re-work the cost calculation in two parts

After a private discussion with Sander I've updated the patch to
calculate the costs in two parts, in a way that's consistent with
the SelectionDAG lowering. I think this is neater and more future
proof since we'll automatically pick up any cost model changes for
either part without having to do any extra work.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2461,6 +2461,25 @@ InstructionCost AArch64TTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
+    InstructionCost Part1 = AArch64TTIImpl::getCastInstrCost(
+        Opcode, LegalTy, Src, CCH, CostKind, I);
+    InstructionCost Part2 = AArch64TTIImpl::getCastInstrCost(
+        Opcode, Dst, LegalTy, CCH, CostKind, I);
```
I think that the CCH for the second getCastInstrCost (from LegalTy -> DstTy) should no longer be `TTI::CastContextHint::Masked`, because it will end up as a regular cast.

david-arm replied:
Good spot! I've changed this to pass in None now, although it didn't make any difference to the cost.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2461,6 +2461,25 @@ InstructionCost AArch64TTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
+    // extend up into two parts:
+    //  1. Perform an extending load or masked load up to the legal type.
+    //  2. Extend the loaded data to the final type.
+    std::pair<InstructionCost, MVT> SrcLT = getTypeLegalizationCost(Src);
```
Should the final cost be multiplied by the legalization cost `SrcLT.first` ?

sdesmalen-arm replied:
You're right, that makes sense. I see you've also added the explicit checks to ensure that SrcTy is promoted and DstTy needs splitting.

david-arm committed:
Pass in a context hint of TTI::CastContextHint::None for second cost

sdesmalen-arm reviewed (approved):
LGTM!
